File: Stack/dynamic/C/src/stack/stack.c
#include "stack.h"

void init_stack(Stack* stack) {
    stack->head = NULL;
}

bool is_empty(Stack* stack) {
    return stack->head == NULL;
}

void push(Stack* stack, Node* node) {
    node->next = stack->head;
    stack->head = node;
}

Node* pop(Stack* stack) {
    if (is_empty(stack)) return NULL;
    Node* node = stack->head;
    stack->head = stack->head->next;
    return node;
}

Node* peek(Stack* stack) {
    if (is_empty(stack)) return NULL;
    return stack->head;
}

void print_stack(Stack* stack) {
    Node* aux_node = stack->head;
    printf("\tHEAD -> ");
    while(aux_node != NULL) {
        printf("%d -> ", aux_node->data);
        aux_node = aux_node->next;
    }
    printf("NULL\n");
}

File: src-ts/Stack/dynamic/Stack.ts
import { Node } from "../../utils/Node";

export default class Stack {
    head: Node | undefined;

    constructor() {
        this.head = undefined;
    }

    is_empty() {
        return !this.head;
    }

    push(node: Node) {
        node.next = this.head;
        this.head = node;
    }

    pop(): Node | undefined {
        if (this.is_empty()) return undefined;
        const node = this.head;
        if (!node) return undefined;
        this.head = node.next;
        return node;
    }

    peek(): Node | undefined {
        return this.head;
    }

    toString(): string {
        let str = "\tHEAD -> ";
        let aux_node = this.head;
        while (aux_node !== undefined) {
            str += `${aux_node.data} -> `
            aux_node = aux_node.next;
        }
        str += "NULL\n";
        return str;
    }
}